pkg/inframetadata/reporter.go

```diff
@@ -0,0 +1,149 @@
+package inframetadata
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/internal/hostmap"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/payload"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes/source"
+)
+
+const (
+	// AttributeDatadogHostUseAsMetadata states if a payload should be used for host metadata.
+	// It overrides the default behavior (see below).
+	AttributeDatadogHostUseAsMetadata = "datadog.host.use_as_metadata"
+
+	// shouldUseByDefault specifies if the payloads should be used by default when the
+	// use_as_metadata resource attribute is missing.
+	shouldUseByDefault = false
+)
+
+type Pusher interface {
+	// Push host metadata to a remote endpoint.
+	// MUST be safe to call concurrently.
+	Push(context.Context, payload.HostMetadata) error
+}
+
+// Reporter of host metadata based on pcommon.Resource payloads.
+type Reporter struct {
+	// logger (sampled) for warnings.
+	logger *zap.Logger
+	// hostMap storing the host metadata.
+	hostMap *hostmap.HostMap
+	// pusher of host metadata.
+	pusher Pusher
+	// closeCh can stop the host metadata reporting.
+	closeCh chan struct{}
+	// ticker for periodic host metadata reporting.
+	ticker *time.Ticker
+}
+
+// Copied over from github.com/open-telemetry/opentelemetry-collector/blob/14c039d/exporter/exporterhelper/queued_retry.go#L269
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+func createSampledLogger(logger *zap.Logger) *zap.Logger {
+	if logger.Core().Enabled(zapcore.DebugLevel) {
+		// Debugging is enabled. Don't do any sampling.
+		return logger
+	}
+
+	// Sample all messages to 1 per 10 seconds initially,
+	// and 1/100 of messages after that.
+	opts := zap.WrapCore(func(core zapcore.Core) zapcore.Core {
+		return zapcore.NewSamplerWithOptions(core, 10*time.Second, 1, 100)
+	})
+	return logger.WithOptions(opts)
+}
+
+// NewReporter creates a new host metadata reporter.
+// The reporter consumes pcommon.Resources through its 'Consume' method and merges them into payload.HostMetadata payloads.
+// It then periodically exporters these functions through the pusher with a specified period.
+func NewReporter(logger *zap.Logger, pusher Pusher, period time.Duration) (*Reporter, error) {
+	hostMap, err := hostmap.New()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build host map: %w", err)
+	}
+
+	return &Reporter{
+		logger:  createSampledLogger(logger),
+		hostMap: hostMap,
+		pusher:  pusher,
+		closeCh: make(chan struct{}),
+		ticker:  time.NewTicker(period),
+	}, nil
+}
+
+// checkResource to see if it should be used by default.
+// A resource is usable if 'AttributeDatadogHostUseAsMetadata' is true or shouldUseByDefault is true.
+func checkResource(res pcommon.Resource) (bool, error) {
```

Comment:
Why not call the function `shouldUse` or even better: `isHostmetadata`? And the variable simply `ok`?